Tools/WebKitTestRunner/wpe/EventSenderProxyClientWPE.cpp

```diff
@@ -303,6 +303,114 @@ void EventSenderProxyClientWPE::keyDown(WKStringRef keyRef, double time, WKEvent
     wpe_event_unref(event);
 }
 
+#if ENABLE(TOUCH_EVENTS)
+void EventSenderProxyClientWPE::addTouchPoint(int x, int y, double)
+{
+    uint32_t id = 0;
+    for (const auto& point : m_touchPoints) {
+        if (point.id == id)
+            id++;
+    }
+    m_touchPoints.append({ id, TouchPoint::State::Down, x, y });
+}
+
+void EventSenderProxyClientWPE::updateTouchPoint(int index, int x, int y, double)
+{
+    ASSERT(index >= 0 && static_cast<size_t>(index) <= m_touchPoints.size());
+
+    auto& point = m_touchPoints[index];
+    point.x = x;
+    point.y = y;
+    point.state = TouchPoint::State::Move;
+}
+
+void EventSenderProxyClientWPE::releaseTouchPoint(int index, double)
+{
+    ASSERT(index >= 0 && static_cast<size_t>(index) <= m_touchPoints.size());
+
+    auto& point = m_touchPoints[index];
+    point.state = TouchPoint::State::Up;
+}
+
+void EventSenderProxyClientWPE::cancelTouchPoint(int index, double)
+{
+    ASSERT(index >= 0 && static_cast<size_t>(index) <= m_touchPoints.size());
+
+    auto& point = m_touchPoints[index];
+    point.state = TouchPoint::State::Cancel;
+}
+
+void EventSenderProxyClientWPE::clearTouchPoints()
+{
+    m_touchPoints.clear();
+}
+
+void EventSenderProxyClientWPE::touchStart(double time)
+{
+    auto* view = WKViewGetView(m_testController.mainWebView()->platformView());
+    for (auto& point : m_touchPoints) {
+        if (point.state != TouchPoint::State::Down)
+            continue;
+
+        point.state = TouchPoint::State::Stationary;
+        auto* event = wpe_event_touch_new(WPE_EVENT_TOUCH_DOWN, view, WPE_INPUT_SOURCE_TOUCHSCREEN, secToMsTimestamp(time), static_cast<WPEModifiers>(m_touchModifiers), point.id, point.x, point.y);
+        wpe_view_event(view, event);
+        wpe_event_unref(event);
```

Comment:
I imagine we don't have some unique pointer support for these classes (yet) to avoid the manual ref/unreffing?

Reply:
We have it in WebKit layer, but not in WTF because wpe platform api is UI process only lib